documentation/docs/explanations/deadlock-analysis.mdx

```diff
@@ -0,0 +1,179 @@
+image: /img/logo/Stronghold_icon.png
+keywords:
+- concurrency
+- deadlock
+- locks
+---
+
+# Deadlock analysis for Stronghold
+
+Current Stronghold use `RwLock` to enable [concurrency](./concurrency.md).
+This approach has the pros of being simple in theory and has fast performance when done right. The downside is that in practice it is usually really difficult to use locks in concurrent programming and you often encounter _data races_ or _deadlocks_ when using them.
+One of the worst case would to have a deadlock in Stronghold which would have the consequences to block Stronghold and not being able to use it.
+To avoid such situation a deadlock analysis has been used which we present here.
+
+## Requirements for multithreaded Strongold
+- Be able to clone the elements parts of the API and use them concurrently. Types reachable through the API that may be cloned and shared between threads 
+  - `Stronghold`
+  - `Client`
+  - `Store`
+- Use `Mutex` or `RWLock`
+- Argument on the impossibility of deadlocks
+
+## Analysis of deadlocks
+This is internal thoughts stemming from experience but does not have peer-reviewed paper backing it.
+
+### Model
+Locks are used to control access to shared memory.
+To simplify the problem we only model mutex so that only a single thread
+can access to a shared memory.
+
+#### Model of function and locks
+Here how we represent a function and locks.
+
+![](./assets/function_model.drawio.png)
+
+- $1$, $2$, $3$ represent locks that protect 3 different and distinct shared memory locations
+- $f_1$ is a function that sequentially takes hold of lock $1$ then $2$ then $3$
+- $f_1$ releases the locks $1$, $2$, $3$ at the end of its computation 
+
+Here we can see that lock $2$ depends on lock $1$ because in $f_1$ you can't take lock $2$ before taking lock $1$. In the same way lock $3$ depends on locks $1$ and $2$.
+
+#### Example: Deadlock with 2 functions
+Here is an example of the most simple possible deadlock.
+
+![](assets/deadlock_2_functions.drawio.png)
+
+- the function $f_1$ wants to take the locks for shared memory $1$ then $2$
+- the function $f_2$ wants to take the locks for shared memory $2$ then $1$
+- on the right you can see that there is a circular dependency of locks between $f_1$ and $f_2$ (you can do a circle by following the arrows of $f_1$ and $f_2$). 
+
+##### Analysis
+
+The graph at the right represents the merged dependency graph of functions $f_1$ and $f_2$. It shows all the dependencies of the locks in $f_1$ and $f_2$.
+The presence of a circular dependency on the merged graph indicates the possibility of a deadlock when executing $f_1$ and $f_2$ concurrently.
+
+##### Example run resulting in a deadlock: 
+1. $f_1$ starts and takes the lock $1$
+2. $f_2$ starts and takes the lock $2$
+3. $f_1$ is stuck because $f_2$ is holding the lock $2$ that it requires
+4. $f_2$ is stuck because $f_2$ is holding the lock $1$ that it requires
```

Comment:
```suggestion
##### Example run resulting in a deadlock 

1. $f_1$ starts and takes the lock $1$.
2. $f_2$ starts and takes the lock $2$.
3. $f_1$ is stuck, because $f_2$ is holding the lock $2$ that it requires.
4. $f_2$ is stuck, because $f_2$ is holding the lock $1$ that it requires.
```